### DISCUSSION TITLE
L2 distance calculation

### DISCUSSION THREAD

**Discussion by BBjie** [milvus-io/milvus#35680] [Q&A and General discussion]
Currently, I'm encountering an issue with getting accurate results back by L2 searching. This is my code 
```
v =  [12.398712,4.6340704,1.9631956,24.852571]  
 
collection.search(
    data=[v],
    anns_field="last_name",
    param={"metric_type": "L2", "params": {"level": 1}},
    output_fields=['truth_row_num',"last_name"],
    limit=2)

```  
and it returns the output 
`data: ['["id: 451951264905112005, distance: 0.001152675598859787, entity: {\'truth_row_num\': 355153, \'last_name\': [12.399810791015625, 4.628012657165527, 1.9622713327407837, 24.93383026123047]}", "id: 451951264905371361, distance: 0.001152675598859787, entity: {\'truth_row_num\': 614509, \'last_name\': [12.399810791015625, 4.628012657165527, 1.9622713327407837, 24.93383026123047]}"]'], cost: 6`

So I go back to the record with the same truth_row_num returned by output and manually check the L2 distance calculation by the following code 
```
import math
def l2_distance(vec1, vec2):
    squared_diffs = [(x - y)**2 for x, y in zip(vec1, vec2)]
    return sum(squared_diffs)
truth_3= [12.398712,4.6340704,1.9631956,24.852571]  
truth_355153= [12.399810791015625, 4.628012657165527, 1.9622713327407837, 24.93383026123047]
distance = l2_distance(truth_3, truth_355153)
print(f"The L2 distance between the vectors is: {distance:.14f}")
``` 
 and the output is 

`The L2 distance between the vectors is: 0.00664182539563
`

which is not aligned with the distance attribute returned by the search query. Is there anything Im doing wrong?

**Comment by xiaofan-luan** [comment]
The distance calculation in Milvus is not the square root version, but rather the squared version.

**Reply by BBjie** [reply]
yes, I manually calculated the L2 distance in the code, which used the squared version as well. but the distance does not equal to what I got back from the search result....

**Comment by yhmo** [comment]
All the vectors are stored as float32 values in milvus server side. Even if you input float64 values from client side, the values are converted to float32 values at client-side and transferred as float32 values by grpc interfaces. When you use "output_fields" to return the vectors, the milvus server returns float32 values to the client, and the python client automatically converted them into float64 values. Although you see the float64 values, actually only 6 ~ 7 digits are precise, the others are **randomly** generated. The real value of 12.399810791015625 is 12.399810 or 12.399811. So, it doesn't make sense to calculate L2 between 12.399810791015625 and 12.399812.

```
truth_3= [12.398712,4.6340704,1.9631956,24.852571]  
truth_355153= [12.399811, 4.6280127, 1.9622713, 24.9338303]
distance = l2_distance(truth_3, truth_355153)
```

To verify the distance, the method should only verify the 6 ~ 7 digits. Such as `math.abs(distance_1 - distance_2) <= 0.000001`

**Reply by BBjie** [reply]
Thank you for the clarification. My question now is: if we want to get accurate search results from the collection, since my vectors are float64, should I convert them all to float32 to get good results, even though it might lose some precision in terms of the data itself?

**Reply by xiaofan-luan** [reply]
> Thank you for the clarification. My question now is: if we want to get accurate search results from the collection, since my vectors are float64, should I convert them all to float32 to get good results, even though it might lose some precision in terms of the data itself?

usually, there is not that much different for Float32 and Float64 result. As far as I know, the result comes from embedding is usually float32 or fp16 format.
